conda/resolve.py

```diff
@@ -497,12 +525,13 @@ def get_reduced_index(self, explicit_specs, sort_by_exactness=True):
 
         def filter_group(_specs):
             # all _specs should be for the same package name
-            name = next(iter(_specs)).name
+            spec = next(iter(_specs))
+            name = spec.name
             group = self.groups.get(name, ())
 
             # implement strict channel priority
             if group and strict_channel_priority and name not in cp_filter_applied:
```

Comment:
On reading and re-reading, this bit is not being handled in a strictly correct way, I think. Since this pass can occur multiple times per package and the key is the package name, but the filter result depends on the spec, only the first spec that gets here will take effect.

The result is that, in the event we have:

- `a` depends on `c==2`
- `a` is in channel-1
- `c 2.0` is only in channel-2
- `c 1.0` is in channel-1
- channel-1 has higher priority than channel-2

The simple cases:

- `conda install a`
- `conda install c==2`

will get `c 2.0` from channel-2, which is the intended change from this PR. Since it was suggested that *ranges* should not slide to lower-priority channels,

- `conda install 'c>=2'` will fail with unsatisfiable

since the exact-version condition is not triggered. Where it gets a little weird is

- `conda install a c>=2`

which fails with unsatisfiable and slightly weird:

```
Package c conflicts for:
a -> c==2
c[version='>=2']
Note that strict channel priority may have removed packages required for satisfiability.
```

This could be surprising because we have `a->c==2.0`, which picks channel-2 for c when only `a` is installed, but since `c>=2` is evaluated first and channel-priority is selected only once per package, it's picking channel-1 for c, then failing unsatisfiably on the `c==2` requirement. So these two requirements are indeed in conflict, but only because they are resulting in different channel selections, which is not communicated in the error message. I'd be okay with that, were it not the case that the exact same requirements entering the solver in a different order would be satisfiable.

I tried a bit to allow the channel-priority pass to run multiple times, which would have to allow picking channel-1 then later picking channel-2 (removing and then re-adding packages in get_reduced_index), but couldn't get it to work since reduce is expected to only decrease the acceptable packages. The result was channel-1 was picked and the solve failed with no candidate for `c==2.0` before the new channel would have been selected after adding the `c==2.0` spec.

I'm not sure it's possible to get this exactly right, or what exactly right really is.